network/p2p/router.go

```diff
@@ -0,0 +1,222 @@
+			zap.Uint32("requestID", requestID),
+			zap.Time("deadline", deadline),
+			zap.Binary("message", request),
+		)
+		return nil
+	}
+
+	return handler.AppRequest(ctx, nodeID, requestID, deadline, parsedMsg)
+}
+
+func (r *Router) AppRequestFailed(_ context.Context, nodeID ids.NodeID, requestID uint32) error {
+	callback, ok := r.clearAppRequest(requestID)
+	if !ok {
+		return ErrUnrequestedResponse
+	}
+
+	callback(nodeID, nil, ErrAppRequestFailed)
+	return nil
+}
+
+func (r *Router) AppResponse(_ context.Context, nodeID ids.NodeID, requestID uint32, response []byte) error {
+	callback, ok := r.clearAppRequest(requestID)
+	if !ok {
+		return ErrUnrequestedResponse
+	}
+
+	callback(nodeID, response, nil)
+	return nil
+}
+
+func (r *Router) AppGossip(ctx context.Context, nodeID ids.NodeID, gossip []byte) error {
+	parsedMsg, handler, ok := r.parse(gossip)
+	if !ok {
+		r.log.Debug("failed to process message",
+			zap.Stringer("messageOp", message.AppGossipOp),
+			zap.Stringer("nodeID", nodeID),
+			zap.Binary("message", gossip),
+		)
+		return nil
+	}
+
+	return handler.AppGossip(ctx, nodeID, parsedMsg)
+}
+
+func (r *Router) CrossChainAppRequest(
+	ctx context.Context,
+	chainID ids.ID,
+	requestID uint32,
+	deadline time.Time,
+	msg []byte,
+) error {
+	parsedMsg, handler, ok := r.parse(msg)
+	if !ok {
+		r.log.Debug("failed to process message",
+			zap.Stringer("messageOp", message.CrossChainAppRequestOp),
+			zap.Stringer("chainID", chainID),
+			zap.Uint32("requestID", requestID),
+			zap.Time("deadline", deadline),
+			zap.Binary("message", msg),
+		)
+		return nil
+	}
+
+	return handler.CrossChainAppRequest(ctx, chainID, requestID, deadline, parsedMsg)
+}
+
+func (r *Router) CrossChainAppRequestFailed(_ context.Context, chainID ids.ID, requestID uint32) error {
+	callback, ok := r.clearCrossChainAppRequest(requestID)
+	if !ok {
+		return ErrUnrequestedResponse
+	}
+
+	callback(chainID, nil, ErrAppRequestFailed)
+	return nil
+}
+
+func (r *Router) CrossChainAppResponse(_ context.Context, chainID ids.ID, requestID uint32, response []byte) error {
+	callback, ok := r.clearCrossChainAppRequest(requestID)
+	if !ok {
+		return ErrUnrequestedResponse
+	}
+
+	callback(chainID, response, nil)
+	return nil
+}
+
+// Parse parses a gossip or request message and maps it to a corresponding
+// handler if present.
+func (r *Router) parse(msg []byte) ([]byte, responder, bool) {
+	handlerID, bytesRead := binary.Uvarint(msg)
+	if bytesRead <= 0 {
+		return nil, responder{}, false
+	}
+
+	r.lock.RLock()
+	defer r.lock.RUnlock()
+
+	handler, ok := r.handlers[handlerID]
+	return msg[bytesRead:], handler, ok
+}
+
+func (r *Router) clearAppRequest(requestID uint32) (AppResponseCallback, bool) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	callback, ok := r.pendingAppRequests[requestID]
+	delete(r.pendingAppRequests, requestID)
+	return callback, ok
+}
+
+func (r *Router) clearCrossChainAppRequest(requestID uint32) (CrossChainAppResponseCallback, bool) {
```

Comment:
We call this with the read lock held though which seems bad 

```suggestion
// Assumes [r.lock] isn't held.
func (r *Router) clearCrossChainAppRequest(requestID uint32) (CrossChainAppResponseCallback, bool) {
```

Reply:
> We call this with the read lock held though which seems bad

No we don't?